### PR TITLE
i18n: centralize the list of supported languages in ./i18n_tool.py

### PR DESCRIPTION
**Pull request by dachary**
## Status

Ready for review

## Description of Changes

Centralize the list of supported languages in ./i18n_tool.py so there is no need to remember to edit multiple files in multiple places when updating or adding a new language.

*  Update the list of supported languages in the documentation. (update-docs)
* Collect supported languages updates/new and commit them with proper commit messages (update-from-weblate)

## Testing

* securedrop/bin/dev-shell ./i18n_tool.py update-docs # verify it does not add any commit
* manually remove one line from docs/includes/l10n.txt
* securedrop/bin/dev-shell ./i18n_tool.py update-docs # verify it adds the line back
* remove **[user]** from .git/config 
* securedrop/bin/dev-shell ./i18n_tool.py --verbose update-docs # verify it fails with a message that suggests the problem may be because ~/.gitconfig does not exist in the docker container
* securedrop/bin/dev-shell ./i18n_tool.py update-from-weblate # verify a few more commits have been added (Norwegian etc.) and check their commit messages

## Deployment

N/A

### If you made changes to documentation:

- [x] Doc linting passed locally

### CODE REVIEW

**Comment by redshiftzero** [comment]
Trivial linting error:

```
docs/includes/l10n.txt:2:Explicit markup ends without a blank line; unexpected unindent.
```

The test failures [here](https://circleci.com/gh/freedomofpress/securedrop/9527) in CI appear to be legitimate and related to the changes in this PR (I also see test failures locally e.g. `tests/test_i18n.py::TestI18N::test_i18n`)

**Comment by dachary** [comment]
@redshiftzero thanks! There still is work to do for this pull request, marking it WIP for now.

**Comment by dachary** [comment]
Interesting case of ci-lint fimeout (first time I see this): https://circleci.com/gh/freedomofpress/securedrop/10404?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Most likely transient.
<pre>
+
test
Spin up Environment00:14
apt-get install -y make00:03
Checkout code00:03
Rebase on-top of latest develop00:00
Setup a remote Docker engine00:02
TAG=${CIRCLE_SHA1} make ci-lint-image25:21
TAG=${CIRCLE_SHA1} make ci-lint10:49 (timed out)

#!/bin/bash -eo pipefail
TAG=${CIRCLE_SHA1} make ci-lint

devops/scripts/dev-shell-ci sudo make --keep-going lint typelint
Run with DOCKER_BUILD_VERBOSE=true for more information
Docker image build in progress  done !
^@^@make -C docs/ clean && sphinx-build -Wn docs/ docs/_build/html
make[1]: Entering directory '/src/docs'
rm -rf _build/*
make[1]: Leaving directory '/src/docs'
Running Sphinx v1.6.3
making output directory...
loading pickled environment... not yet created
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 66 source files that are out of date
updating environment: 66 added, 0 changed, 0 removed
reading sources... [100%] yubikey_setup

looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] yubikey_setup

generating indices... genindex
writing additional pages... search
copying images... [100%] development/../images/weblate/review1.png

copying downloadable files... [100%] images/labels/journalist_workstation.png

copying static files... done
copying extra files... done
dumping search index in English (code: en) ... done
dumping object inventory... done
build succeeded.
cd securedrop && make lint
make[1]: Entering directory '/src/securedrop'
find . -name '*.py' | xargs pylint --reports=no --errors-only --disable=import-error
Using config file /src/securedrop/pylintrc
make[1]: Leaving directory '/src/securedrop'
flake8 --exclude='config.py,.venv/'
html_lint.py --printfilename --disable=optional_tag,extra_whitespace,indentation,names \
	securedrop/source_templates/*.html securedrop/journalist_templates/*.html
/src/testinfra/vars/mon-prod.yml
  31:201    warning  line too long (206 > 200 characters)  (line-length)

/src/testinfra/vars/app-staging.yml
  110:201   warning  line too long (206 > 200 characters)  (line-length)
  117:201   warning  line too long (212 > 200 characters)  (line-length)

/src/testinfra/vars/app-prod.yml
  61:201    warning  line too long (206 > 200 characters)  (line-length)
  68:201    warning  line too long (212 > 200 characters)  (line-length)

/src/molecule/aws/aws-launch.yml
  29:22     warning  too many spaces after colon  (colons)

Makefile:28: recipe for target 'ci-lint' failed
make: *** [ci-lint] Terminated
Too long with no output (exceeded 10m0s)
</pre>

**Comment by dachary** [comment]
Another timeout, in a place that does not look like it should timeout. https://circleci.com/gh/freedomofpress/securedrop/10417?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

<pre>
    TASK [Start Tor docker proxy for use in functional app tests.] *****************
    Friday 16 March 2018  21:22:19 +0000 (0:00:00.047)       0:00:01.047 **********
    changed: [localhost]
    
    TASK [Start functional docker container] ***************************************
    Friday 16 March 2018  21:25:52 +0000 (0:03:33.920)       0:03:34.968 **********
Makefile:20: recipe for target 'ci-go' failed
make: *** [ci-go] Terminated
Too long with no output (exceeded 10m0s)
</pre>

**Comment by dachary** [comment]
@redshiftzero this is ready for review now. There still is a lot to do to automate i18n maintenance but... that's a step forward. I also think this manual step is the most error prone: I always feel nervous when doing it ;-)

File: docs/development/i18n.rst [hunk 1]
```diff
@@ -161,23 +161,11 @@ to the ``develop`` branch via pull requests for merge on a regular basis.
 
       $ git clone https://github.com/freedomofpress/securedrop
       $ cd securedrop
-      $ git remote add lab http://lab.securedrop.club/bot/securedrop/tree/i18n
-      $ git fetch lab
       $ git checkout -b wip-i18n origin/develop
```
Comment (redshiftzero) [inline docs/development/i18n.rst]: Before we sync from weblate, what about adding a note (maybe in a subsection "Adding a New Supported Language" or similar) which basically says:

1. Modify `SUPPORTED_LANGUAGES` in `securedrop/i18n_tool.py` to add the new language
2. Run `securedrop/bin/dev-shell ./i18n_tool.py update-docs`

File: docs/development/i18n.rst [hunk 2]
```diff
@@ -161,23 +161,11 @@ to the ``develop`` branch via pull requests for merge on a regular basis.
...
-      $ sd="nl fr de_DE nb_NO pt_BR es_ES zh_Hant tr it ar"
-      $ for l in $sm ; do \
-           git add securedrop/translations/$l/LC_MESSAGES/messages.po ; \
-        done
-      $ for l in $sd ; do \
-           git checkout lab/i18n -- \
-               install_files/ansible-base/roles/tails-config/templates/$l.po ; \
-           git add install_files/ansible-base/roles/tails-config/templates/$l.po ; \
-        done
+      $ securedrop/bin/dev-shell ./i18n_tool.py --verbose update-from-weblate
```
[the remark this answers is not in the document]
Comment (dachary) [reply docs/development/i18n.rst]: good catch! Fixed :-)

File: docs/development/i18n.rst (hunk 1 above)
Comment (dachary) [reply docs/development/i18n.rst]: Ooops, missed this part. Added a paragraph in the documentation.

**Comment by dachary** [comment]
@redshiftzero rebased & fixed & repushed. The existing commits were not modified to facilitate review. Two commits were added:

* i18n_tool: verify user.email & user.name are set before commit
* SQUASH: i18n_tool: add the update-docs subcommand

Thanks for your review :-)

**Comment by redshiftzero** [comment]
Ah, CI is failing because bandit discovered the use of `shell=True` in a `subprocess` call. Since this is in code that will not be executed on the application server, I think it is reasonable to add `#nosec` (@emkll let us know if you disagree)

**Comment by dachary** [comment]
fixed & repushed

**Comment by dachary** [comment]
rebased, resolved a trivial conflict due to awesome @kushaldas IO changes :tada: and repushed. Nothing changed otherwise.

**Comment by dachary** [comment]
@emkll would you have time to check the **# nosec** is not too risky?

File: securedrop/i18n_tool.py [hunk 3]
```diff
@@ -57,196 +59,371 @@ def sh(command, input=None):
...
+        return True
+
+    def update_docs(self, args):
+        l10n_content = '.. GENERATED BY i18n_tool.py DO NOT EDIT:\n\n'
+        for (code, info) in sorted(I18NTool.SUPPORTED_LANGUAGES.items()):
+            l10n_content += '* ' + info['name'] + ' (' + code + ')\n'
+        includes = join(args.documentation_dir, 'includes')
+        l10n_txt = join(includes, 'l10n.txt')
+        open(l10n_txt, 'w').write(l10n_content)
+        self.require_git_email_name(includes)
```
Comment (redshiftzero) [inline securedrop/i18n_tool.py]: Is `includes` the right directory to be passed as `git_dir` to `require_git_email_name`? I'm getting an exception when I try to run this utility, even with my name and email defined in a repo-level `.gitconfig`.

Comment (dachary) [reply securedrop/i18n_tool.py]: it does not matter if the command is not run at the root of the git repository, the important thing is that it runs within a git repository

**Comment by dachary** [comment]
@redshiftzero I updated the test instructions in the PR message to verify the corner case you just ran into displays an error message that suggests the solution. Thanks for your patience!
